packit_service/service/api/webhooks.py

```diff
@@ -206,28 +208,95 @@ def post(self):
+                f"the Secret Token `{token}` to authenticate requests coming to Packit.\n\n"
+                f"Packit needs rights to set commit status to merge requests, Please grant\n"
+                f"[{packit_user}](https://gitlab.com/{packit_user}) `admin` permissions\n"
+                f"on the {namespace}/{repo_name} project. You can add the rights by clicking\n"
+                f"[here]({http_url}/-/project_members).",
+                private=True,
+            )
+
+            ProjectTokenModel.create(
+                namespace=namespace,
+                repo_name=repo_name,
+                project_url=http_url,
+                issue_created=True,
+            )
+
+    def validate_token(self):
         """
         https://docs.gitlab.com/ee/user/project/integrations/webhooks.html#secret-token
         """
         if "X-Gitlab-Token" not in request.headers:
             if config.validate_webhooks:
                 msg = "X-Gitlab-Token not in request.headers"
-                logger.warning(msg)
-                raise ValidationFailed(msg)
+                logger.info(msg)
+                self.create_confidential_issue_with_token()
+                return
             else:
                 # don't validate signatures when testing locally
                 logger.debug("Ain't validating token.")
                 return
 
         token = request.headers["X-Gitlab-Token"]
 
-        # Find a better solution
-        if token not in config.gitlab_webhook_tokens:
-            raise ValidationFailed("Payload token validation failed.")
+        if token in config.gitlab_webhook_tokens:
+            logger.debug("Deprecation Warning: Old Gitlab tokens used.")
+            return
+
+        gitlab_token_secret = config.gitlab_token_secret
+        if not gitlab_token_secret:
+            msg = "'gitlab_token_secret' not specified in the config."
+            logger.error(msg)
+            raise ValidationFailed(msg)
 
-        logger.debug("Payload token is OK.")
+        msg_failed_validation = "Signature of the payload token is not valid."
+        msg_failed_signature = "Payload token does not match the project."
+
+        try:
+            token_decoded = jwt.decode(
+                token, config.gitlab_token_secret, algorithms=["HS256"]
+            )
+        except jwt.exceptions.InvalidSignatureError:
+            logger.warning(msg_failed_signature)
+            raise ValidationFailed(msg_failed_signature)
+
+        project_data = request.json["project"]
+        namespace, repo_name = project_data["path_with_namespace"].split("/")
+        token_namespace = token_decoded["namespace"]
+        token_repo_name = token_decoded["repo_name"]
+
+        if (token_namespace, token_repo_name) == (namespace, repo_name):
+            logger.debug("Payload signature is OK.")
+        else:
+            logger.warning(msg_failed_validation)
+            raise ValidationFailed(msg_failed_validation)
```

Comment:
If I can suggest another nit-pick change..;)

```suggestion
        try:
            token_decoded = jwt.decode(
                token, config.gitlab_token_secret, algorithms=["HS256"]
            )
        except jwt.exceptions.InvalidSignatureError:
            msg_failed_signature = "Payload token does not match the project."
            logger.warning(msg_failed_signature)
            raise ValidationFailed(msg_failed_signature)

        project_data = request.json["project"]
        namespace, repo_name = project_data["path_with_namespace"].split("/")
        token_namespace = token_decoded["namespace"]
        token_repo_name = token_decoded["repo_name"]

        if (token_namespace, token_repo_name) == (namespace, repo_name):
            logger.debug("Payload signature is OK.")
        else:
            msg_failed_validation = "Signature of the payload token is not valid."
            logger.warning(msg_failed_validation)
            raise ValidationFailed(msg_failed_validation)
```